File: LeetCode/Easy/0110-balanced-binary-tree/0110-balanced-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
    _, bal := dfs_helper(root)
    return bal
}

func dfs_helper(root *TreeNode) (int,bool){
    if root == nil{
        return 0, true
    }

    leftHeight, Lbalance := dfs_helper(root.Left)
    rightHeight, Rbalance := dfs_helper(root.Right)

    isNodeBalance := Lbalance && Rbalance && (math.Abs(float64(leftHeight-rightHeight)) <=1)
    height := 1 + max(leftHeight, rightHeight)

    if isNodeBalance{
        return height, true
    }else{
        return -1, false
    }
}

func max(a,b int) int{
    if a > b {
        return a
    } 
    return b
}

File: LeetCode/Easy/0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        pair<int, bool> res = dfs(root);
        return res.second;
    }
    pair<int, bool> dfs(TreeNode* root) {
        if (!root)
            return {0, true};
        pair<int, bool> lr = dfs(root->left);
        pair<int, bool> rr = dfs(root->right);

        int height = 1 + max(lr.first, rr.first);
        bool balanced =
            lr.second && rr.second && (abs(lr.first - rr.first) <= 1);
        return {height, balanced};
    }
};